core/trino-main/src/main/java/io/trino/execution/scheduler/ScaledWriterScheduler.java

```diff
@@ -105,7 +108,11 @@ private int getNewTaskCount()
         // When there is a big data skewness, there could be a bottleneck due to the skewed workers even if most of the workers are not over-utilized.
         // Check both, weighted output buffer over-utilization rate and average output buffer over-utilization rate, in case when there are many over-utilized small tasks
         // due to fewer not-over-utilized big skewed tasks.
-        if (isSourceTasksBufferFull() && isWriteThroughputSufficient() && scheduledNodes.size() < maxWriterNodeCount) {
+        if (isSourceTasksBufferFull()
+                // Do not scale up if the write throughput is not sufficient or the memory is not fully utilized
+                // across available tasks.
+                && (isWriteThroughputSufficient() || isMaxMemoryUtilizedAcrossAvailableTasks())
```

Comment:
It's not clear to me what we're trying to achieve through this change. It looks like it will make scale up of tasks faster when per-node memory usage is high. But the link between adding writer tasks on new nodes and reducing per-node memory usage on the existing  nodes is not obvious. The already active orc/parquet writers will buffer a fixed amount of data in memory before flushing to disk no matter what we do here.